internal/graph/taskrunner.go

```diff
@@ -0,0 +1,134 @@
+package graph
+
+import (
+	"context"
+	"sync"
+)
+
+type token struct{}
+
+// TaskRunner is a helper which runs a series of scheduled tasks against a defined
+// limit of goroutines.
+type TaskRunner struct {
+	// ctx holds the context given to the task runner and annotated with the cancel
+	// function.
+	ctx    context.Context
+	cancel func()
+
+	// err holds the error returned by any task, if any. If the context is canceled,
+	// this err will hold the cancelation error.
+	errOnce sync.Once
+	err     error
```

Comment:
nit: since you are a comment over the `err` field, make the comment on top of it
```suggestion
	// err holds the error returned by any task, if any. If the context is canceled,
	// this err will hold the cancelation error.
	err     error
	errOnce sync.Once
```